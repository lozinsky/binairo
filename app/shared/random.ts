import { type RandomGenerator, uniformIntDistribution, xoroshiro128plus } from 'pure-rand';

import { expectToBeDefined } from '~/shared/expect';

export class Random {
  get state() {
    return this.#generator.getState();
  }

  #generator: RandomGenerator;

  constructor(generator: RandomGenerator) {
    this.#generator = generator;
  }

  static create() {
    return new this(xoroshiro128plus(Date.now() ^ (Math.random() * 0x100000000)));
  }

  static from(state: readonly number[]) {
    return new this(xoroshiro128plus.fromState(state));
  }

  static stable() {
    return new this(xoroshiro128plus(0));
  }

  next(from: number, to: number) {
    const [value, generator] = uniformIntDistribution(from, to, this.#generator);

    this.#generator = generator;

    return value;
  }
}

export function sample<T>(target: readonly T[], random: Random) {
  return expectToBeDefined(target[random.next(0, target.length - 1)]);
}

export function shuffle<T>(target: readonly T[], random: Random) {
  const items = [...target];

  for (let prevIndex = 0; prevIndex < items.length; prevIndex++) {
    const nextIndex = random.next(prevIndex, items.length - 1);
    const prevItem = expectToBeDefined(items[prevIndex]);
    const nextItem = expectToBeDefined(items[nextIndex]);

    items[prevIndex] = nextItem;
    items[nextIndex] = prevItem;
  }

  return items;
}
