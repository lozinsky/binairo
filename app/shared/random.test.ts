import { expect, test } from 'vitest';

import { Random, sample, shuffle } from './random';

test('returns random with stable state', () => {
  expect(Random.stable().state).toMatchInlineSnapshot(`
    [
      -1,
      -1,
      0,
      0,
    ]
  `);
});

test.each([
  [0, 100],
  [-100, 0],
  [-100, 100],
])('returns random number', (from, to) => {
  expect(Random.stable().next(from, to)).toMatchSnapshot();
});

test.each([[[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]], [[10, 9, 8, 7, 6, 5, 4, 3, 2, 1]]])(
  'returns shuffled array',
  (target) => {
    expect(shuffle(target, Random.stable())).toMatchSnapshot();
  },
);

test.each([[[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]], [[10, 9, 8, 7, 6, 5, 4, 3, 2, 1]]])(
  'returns random array item',
  (target) => {
    expect(sample(target, Random.stable())).toMatchSnapshot();
  },
);
