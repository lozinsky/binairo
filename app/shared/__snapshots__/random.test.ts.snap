// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`returns random array item 1`] = `8`;

exports[`returns random array item 2`] = `3`;

exports[`returns random number 1`] = `33`;

exports[`returns random number 2`] = `-67`;

exports[`returns random number 3`] = `-51`;

exports[`returns shuffled array 1`] = `
[
  8,
  5,
  3,
  4,
  10,
  6,
  9,
  2,
  7,
  1,
]
`;

exports[`returns shuffled array 2`] = `
[
  3,
  6,
  8,
  7,
  1,
  5,
  2,
  9,
  4,
  10,
]
`;
